### ISSUE TITLE
Implement smt_cap_reveal_operation

### ISSUE THREAD

**Issue by theoreticalbts** [steemit/steem#1511]
Implement all semantics of the `smt_cap_reveal_operation`.

- Revealed cap must match existing cap definition
- Must be cryptographically checked against revealed nonce (use `smt_revealed_cap::validate()` against `smt_cap_commitment` stored in the database)
- Unit tests

**Comment by youkaicountry** [comment]
The evaluator for this is currently a stub.

**Comment by Kiwonik** [comment]
@theoreticalbts @youkaicountry
1. `smt_revealed_cap::validate()` mentioned in OP is not implemented yet. Is its implementation part of this issue (I don't think so)?
2. `smt_cap_reveal_operation` is valid only when issued in specific time period, i.e. after setup phase is concluded but before fail phase / launch happens. Thus the first thing to do for the evaluator is to check the smt_token_object's phase field to be at least `setup_completed`. This condition is never met, because `smt_setup_operation` is not yet implemented. Therefore I mark this issue as blocked until https://github.com/steemit/steem/issues/1508 is finished and merged.

**Comment by theoreticalbts** [comment]
> `smt_cap_reveal_operation` is valid only when issued in specific time period, i.e. after setup phase is concluded but before fail phase / launch happens

This is not true.  The `smt_cap_reveal_operation` can occur in the same transaction as the `smt_setup_operation` (which contains `smt_cap_commitment` members in `initial_generation_policy`).  Or it can occur at any later time, even after `announced_launch_time`.

- If the caps are revealed before `announced_launch_time`, then at `announced_launch_time`, you check whether there is enough STEEM or not.  If yes, you launch; if no, you start allowing refunds.
- If `announced_launch_time` has arrived and the caps aren't known, then you start allowing refunds.
- If the caps become known after `announced_launch_time` has passed, you check whether there is enough STEEM or not (since you started allowing refunds at `announced_launch_time`, some people might have gotten refunds).  If yes, you launch; if no, you continue allowing refunds.  Since existing contributions can be refunded, but no new contributions can be made, you never have to worry about whether there is enough STEEM after checking it once.

**Comment by Kiwonik** [comment]
@theoreticalbts Your answer does not contradict my statement though you seem to imply so. Let me rephrase my statement then:
1. Successfull evaluation of `smt_setup_operation` **must precede** evaluation of `smt_cap_reveal_operation` if the latter is to succeed. The former is the beginning of the time window in which cap can be revealed - no matter if the two operations happen to occur in the same transaction/block or not. Hence my statement "_after setup phase is concluded_". 
This is consistent with smt-whitepaper ICO timeline diagram https://github.com/steemit/smt-whitepaper/pull/172 
2. Evaluation of `smt_cap_reveal_operation` must fail when either **launch already happened** or ICO is in **failed phase** (because announced_launch_time expired without launch). It just makes no sense to reveal cap after one of the two happens. Hence my statement "_before fail phase / launch happens_" Again, this is consistent with smt-whitepaper ICO timeline diagram https://github.com/steemit/smt-whitepaper/pull/172 

If you insist that cap can be revealed 
a) before smt_setup_operation concludes setup phase of ICO or 
b) after successfull launch of SMT or 
c) during the failed phase (as defined on timeline) 
then please explain the rationale to allow that, as it needs to be put in whitepaper and will result in timeline diagram alterations.

**Comment by Kiwonik** [comment]
@youkaicountry You're probably well aware of it, but I wanted to reiterate from my earlier comment above that except #2152 (that you just created) also #1508 is a prerequisite to this issue.

**Comment by vogel76** [comment]
@youkaicountry I see this operation has missing smt field (stores only control_account). So it shall be supplemented by asset_symbol_type smt field to make possible proper identification of SMT we want to reveal cap for.
As a side note, I propose to make smt_base_operation class, which holds:
- control_account and smt members
- implements validate method verifying these 2 fields.
and avoid redundancy in the code.

**Comment by vogel76** [comment]
@youkaicountry Also only one-time execution shall be allowed for this operation. If given cap has been revealed already, evaluation shall fail to avoid blockchain spamming with large number of the same (correct) operations.
The whitepaper does not contain any description of behavior - for sure we need to know also other ops which are designed as one-shot entities.

**Comment by youkaicountry** [comment]
@vogel76 I am in favor of having a base operation class for SMT ops. 
Regarding the 'one-shot' entities:
The WP definitely needs audited for these cases.

**Comment by mvandeberg** [comment]
> evaluation shall fail to avoid blockchain spamming with large number of the same (correct) operations.

I am in favor of of having these operations validate only once because it makes the most sense logically. You're not wrong in your conclusion, but your reasoning is invalid. The argument that we should do it to not fill up the blockchain does not make sense. We already have custom ops that are not rejected by consensus and can fill up the blockchain. Adding individual op rate limits are not an effective means by which to protect the blockchain. I only bring this up because in my experience it takes many reminders and time to start thinking this way.

Again, this is the correct way to implement this op because there is no need to reveal the cap a second time.

**Comment by Kiwonik** [comment]
@youkaicountry @theoreticalbts In SMT whitepaper conditional phrases are used related to min/hard caps "ICOs may specify ...". I find these phrases misleading, because the construction of `smt_capped_generation_policy` forces ICO creator to specify some values in each cap's commitment. I believe these phrases should be changed to "ICOs have to specify ..." and safe default values should be explicitly mentioned (0 for min cap and STEEM_MAX_SHARE_SUPPLY+1 for hard cap).

**Comment by vogel76** [comment]
@youkaicountry @theoreticalbts Inside method `void smt_capped_generation_policy::validate()const` are conditions completely not addressed in the whitepaper. Also if such additional restrictions shall be applied, cap default values must be assigned accordingly.

**Comment by theoreticalbts** [comment]
The contents of this method aren't mentioned in the whitepaper, because they are implementation details.  Maybe they should be mentioned in the whitepaper.  I'm not sure.

The idea here is that the SMT creator chooses a value of the cap, let's call that `x`.  There are reasons some creators *may not* want to reveal the value of `x` while the ICO is in progress.  Such a creator has three options:

- (Option 1) Reveal the value of `x` immediately.
- (Option 2) Reveal some bounds 'a' and `b`, so that participants know `a <= x <= b`, but do not know the value of `x`.
- (Option 3) Do not reveal any bounds, so the value of `x` can be anything.

*From the creator's point of view* those are the options, and that's what the whitepaper says.

Now *from the implementor's point of view*, it looks like those are *three cases that must be coded separately*.  It results in simpler code (less branching paths) if all three cases are collapsed into one.

In case of option (1), we set `a` and `b` to be equal to each other.  So option (1) is a special case of option (2).  The nonce is irrelevant in this case, since the value is not secret.  Since the nonce is irrelevant, I force it to be zero.  If you know the nonce (zero) and you know `a` and `b`, you have all the information necessary at this point to evaluate the hash `H(x, nonce)`.  So instead of checking it later, we can check it now.

What about option (3)?  In this case "`x` can be anything," but what, precisely, is meant by the word "anything"?

- `x` must be a non-negative integer number of satoshis.  Allowing negative numbers, or data types other than integers, doesn't make sense.
- Does `x == 0` make sense?  I don't think it does.  Maybe there's a use case for it, but I don't see it, and I don't think it's important.  So `x` must be at least `1`.
- Unless you want to use arbitrary-precision integers to track balances, `x` can be at most `std::numeric_limits< uint64_t >::max()`.
- Formatting code in UI's, and possibly some calculations in `steemd`, assume that total supply is at most `STEEM_MAX_SHARE_SUPPLY`.  As this code will also handle SMT's, I currently envision each SMT can have at most `STEEM_MAX_SHARE_SUPPLY`.

So even though in option (3), the value of `x` can be "anything", in reality it is still restricted to `1 <= x <= STEEM_MAX_SHARE_SUPPLY`.  So option (3) is *also* a special case of option (2), where `a = 1` and `b = STEEM_MAX_SHARE_SUPPLY`.

I think this explains all the `FC_ASSERT()` statements except for one, namely `FC_ASSERT( lower_bound <= upper_bound )`.  The idea is that some `x` exists, such that `a <= x`, and `x <= b`, and `H(x, nonce) == this->hash`.  It would be nice to reject cases where no such `x` exists.

Of course, in general, we cannot always reject these cases (unless we have the capability to brute-force `nonce` or crack the hash function).

But if there are specific cases where no such `x` exists, that are easy to detect, we should reject them immediately.  One such case is when `a > b`.  If `a > b`, then we have `x >= a > b >= x` which is a contradiction (no value of `x` satisfies `x > x`).
